scripts/ssh

```diff
@@ -2,6 +2,12 @@
 . "$(dirname "$0")/lib/init"
 
 main() {
+	if [[ $(dd if=secrets/ssh/id_ed25519 bs=1 skip=1 | head -c 8) == "GITCRYPT" ]]; then
```

Comment:
This process should already be automated. Did you `git-crypt init`?

Reply:
It isn't already automated - if you `git-crypt lock` it will not automatically unlock right now on the current version - and yes

Reply:
That should be part of the `shellHook` then, rather than part of a one-off SSH script. Or ideally, the user shouldn't `git-crypt lock` at all.

Reply:
that continues the nix debate from #22 - which my point there was more that it doesn't make sense to `nix-shell` every time just to SSH. if you're gonna be terraforming something then yeah sure, but to just check on something quick via ssh isn't exactly a wonderful `nix-shell` use case. it's variable but it takes anywhere from 5 to 50 seconds on my laptop (NVMe, 4900hs, and connected to a fast and reliable network) to open a nix-shell when the packages have already been downloaded, for a quick ssh that's not ideal. as for not doing `git-crypt lock`, sure, although we should probably add another line to the README anyways saying to do `git-crypt unlock` (`git-crypt init` does not unlock the repo for me, only `unlock` does - `init` just prints a message saying the repo already uses git-crypt)

Reply:
> it doesn't make sense to `nix-shell` every time just to SSH

The way around this should be to use [direnv](https://direnv.net/), which automatically drops you into the Nix shell. In the future, more things may be required in the development environment to do anything at all, so I'm personally against supporting another way of interacting with this repo.

direnv also caches the shell, so entering the shell takes no time with it.

> although we should probably add another line to the README anyways saying to do `git-crypt unlock` (`git-crypt init` does not unlock the repo for me, only `unlock` does - `init` just prints a message saying the repo already uses git-crypt)

This is certainly the odd behavior. A note in the README to run `git-crypt unlock` *if the files aren't already unlocked* would be a good idea, but `git-crypt init`, if played nicely, should already unlock the files automatically.

> [!TIP]
> You can use the `[!TIP]` syntax to add a Tip box into the README for this.

Reply:
> The way around this should be to use [direnv](https://direnv.net/)

Now we're making people further alter their existing development environment, not saying we can't but for the use case the proposed solution works, and decently well at that... to be clear I'm not *opposed* to `direnv`, just feel like it's another layer of complexity when this is a pretty simple addition

> This is certainly the odd behavior. A note in the README to run git-crypt unlock if the files aren't already unlocked would be a good idea, but git-crypt init, if played nicely, should already unlock the files automatically.

Want me to PR in that tip? Wouldn't resolve this PR, just would probably be a good thing to add anyways. (Also I do know about tips)... would it be better to classify this as a tip or as info though?

Reply:
> when this is a pretty simple addition

This is true! It is a very simple addition, but said simple addition comes with the implication that we're either supporting a new way of interacting with the repository, or we're agreeing that it's a half-supported way to half-interact with it, which might very well be dropped in the near future (or even tomorrow for that matter).

With that, the question then becomes: is it worth it to support a new way of interacting with the repository when it comes with these costs? Is using Nix shell and/or Direnv that hard that it would justify the new way?

> would it be better to classify this as a tip or as info though?

Either will work fine.

Reply:
> said simple addition comes with the implication that we're either supporting a new way of interacting with the repository, or we're agreeing that it's a half-supported way to half-interact with it, which might very well be dropped in the near future (or even tomorrow for that matter)

I believe the latter would be true. There is a potential compromise solution: add `chmod` to the `shellHook`, and have instructions for non-nix-shell users. For instance, I currently have have the following exports in my zshrc, along with functions for faster ssh to 306/cirno and my home entry point:

```sh
# acmCSUF environment
set -o allexport
source ~/dev/acm-aws/.env
set +o allexport

function to306() {
  # ...
}

function tocirno() {
  # ...
}

function tosappho() {
  ssh root@10.1.0.1
}
```

> With that, the question then becomes: is it worth it to support a new way of interacting with the repository when it comes with these costs? Is using Nix shell and/or Direnv that hard that it would justify the new way?

Is Direnv selectively enable-able? And is there a way we could automate Direnv setup in a non-destructive way for users if so? That's the only way I can see this being practical - I'd be fine using Direnv *within the scope of acm-aws only*, but due to hacky workarounds that may be present in other environments I would not want to use Direnv elsewhere. Ideally I also want to not create an even higher entry barrier for people - thin out the learning curve a bit, for instance.

Will it even be possible to ssh without nix-shell post-#21? I honestly don't know what the implications of transitioning to flakes are.

> Either will work fine.

Alright - think if we move forward with that README addition, my recommendation would be info over tip; seems slightly closer to the intended point, plus (and this is purely subjective) info looks nicer than tip. 

Also, on both my laptop (Fedora 39) and desktop (Arch), this is what `init` gets me (no decrypt):
```sh
❯ git-crypt init
Error: this repository has already been initialized with git-crypt.
```

In fact, for our use case - "decrypt this repo using the in-repo GPG-encrypted key" - the helppage says we should be using `unlock`, *not* `init`:

```
❯ git-crypt
Usage: git-crypt COMMAND [ARGS ...]

Common commands:
  init                 generate a key and prepare repo to use git-crypt
  status               display which files are encrypted
  lock                 de-configure git-crypt and re-encrypt files in work tree

GPG commands:
  add-gpg-user USERID  add the user with the given GPG user ID as a collaborator
  unlock               decrypt this repo using the in-repo GPG-encrypted key
```

I wonder if this could come from legacy git-crypt - since `git-crypt init KEYFILE` is listed as an alias for `git-crypt unlock KEYFILE`...

Reply:
> Will it even be possible to ssh without nix-shell post-#21? I honestly don't know what the implications of transitioning to flakes are.
> 
> >

These are all great questions. I don't think I can answer them until the PR is finalized, and currently I'm just working on it. Everything could still be changed.

Issue #20 lists the upsides of migrating to Flakes, but Nix 4 (not real yet) is supposed to be Flakes-only, and a lot of core Nix people are adamant on encouraging use of Flakes, so it's best for us to try and heed that advice for maintainability.



> Alright - think if we move forward with that README addition, my recommendation would be info over tip; seems slightly closer to the intended point, plus (and this is purely subjective) info looks nicer than tip.

This sounds good.

> In fact, for our use case - "decrypt this repo using the in-repo GPG-encrypted key" - the helppage says we should be using `unlock`, _not_ `init`

This makes sense. `git-crypt`'s usage means we shouldn't even be worrying about having to decrypt things, because the git hooks are supposed to handle it. But the hooks are sometimes pretty fragile, e.g. they might not run until you do `git status` (unconfirmed), so this clarification is valid.

I believe `git-crypt init` is supposed to set up the git hooks, so it's weird that it would warn that it's "already been initialized" on the first run.

Reply:
> These are all great questions. I don't think I can answer them until the PR is finalized, and currently I'm just working on it. Everything could still be changed.
> Issue https://github.com/acmcsufoss/acm-aws/issues/20 lists the upsides of migrating to Flakes, but Nix 4 (not real yet) is supposed to be Flakes-only, and a lot of core Nix people are adamant on encouraging use of Flakes, so it's best for us to try and heed that advice for maintainability.

Alright, sounds good - how about we put this PR on hold until #21 is done?

> I believe git-crypt init is supposed to set up the git hooks, so it's weird that it would warn that it's "already been initialized" on the first run. If the hooks are missing, that would be pretty dangerous, since someone might try to push unencrypted secrets over.

From my (albeit limited) experience the hooks appear to already be set in just fine - without doing `git-crypt init` it doesn't appear to push unencrypted secrets, but 🤷🏼‍♀️

Reply:
> Alright, sounds good - how about we put this PR on hold until #21 is done?

SGTM.